vms/avm/vm.go

```diff
@@ -495,26 +496,34 @@ func (vm *VM) GetUTXOs(
 		} else if comp == 0 {
 			start = startUTXOID
 		}
-		utxoIDs, err := vm.state.Funds(addr.Bytes(), start, limit) // Get UTXOs associated with [addr]
-		if err != nil {
-			return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXOs for address %s", addr)
-		}
-		for _, utxoID := range utxoIDs {
-			if seen.Contains(utxoID) { // Already have this UTXO in the list
-				continue
-			}
-			utxo, err := vm.state.UTXO(utxoID)
+		for {
+			currLimit := limit
+			utxoIDs, err := vm.state.Funds(addr.Bytes(), start, limit) // Get UTXOs associated with [addr]
 			if err != nil {
-				return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXO %s: %w", utxoID, err)
+				return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXOs for address %s", addr)
 			}
-			utxos = append(utxos, utxo)
-			seen.Add(utxoID)
-			lastAddr = addr
-			lastIndex = utxoID
-			limit--
-			if limit <= 0 {
-				break // Found [limit] utxos; stop.
+			for _, utxoID := range utxoIDs {
+				if seen.Contains(utxoID) { // Already have this UTXO in the list
+					continue
+				}
+				utxo, err := vm.state.UTXO(utxoID)
+				if err != nil {
+					return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXO %s: %w", utxoID, err)
+				}
+				utxos = append(utxos, utxo)
+				seen.Add(utxoID)
+				lastAddr = addr
+				lastIndex = utxoID
+				currLimit--
+				if currLimit <= 0 {
+					break // Found [currLimit] utxos; stop current set.
+				}
+			}
+
+			if paginate || utxoIDs == nil || len(utxoIDs) == 0 { // Avoiding a downstream make([]) making this an infinite loop
```

Comment:
The `utxoIDs == nil` case is caught by `len(utxoIDs) == 0` since `len([a nil slice]) == 0` so we don't need `utxoIDs == nil` here

Reply:
Yeah - I suppose I'm just extra weary when it comes to doing nil checks on slices.
If by any chance the slice is `make`d with a size it will return `len(utxoIDs) == 0 => false`

I just double checked there https://play.golang.org/p/HePW4N-8ncX

Funny enough this seems to have changed (which makes sense) from a few years back https://www.pixelstech.net/article/1539870875-Empty-slice-vs-nil-slice-in-GoLang

If that's ok, I'd prefer to keep this extra validation as all it takes is someone changing the downstream IDs method to blow up here.

Reply:
Fixed 👍